tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -46,52 +47,55 @@ parseByteString language bytestring = withParser language $ \ parser -> withPars
   else
     withRootNode treePtr $ \ rootPtr ->
       withCursor (castPtr rootPtr) $ \ cursor ->
-        runM (runFail (runReader cursor (runReader bytestring unmarshalNode)))
+        runM (runFail (runReader cursor (runReader bytestring (peekNode >>= unmarshalNodes . maybeToList))))
 
 -- | Unmarshal is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
 --
 --   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'unmarshalNode' providing that they have a suitable 'Generic' instance.
 class Unmarshal a where
-  unmarshalNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
-  default unmarshalNode :: (MonadFail m, Carrier sig m, GUnmarshal (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
-  unmarshalNode = to <$> gunmarshalNode
 
-  unmarshalEmpty :: MonadFail m => m a
-  unmarshalEmpty = fail "expected a node but didn't get one"
+  unmarshalNodes :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
+  default unmarshalNodes :: (MonadFail m, Carrier sig m, GUnmarshal (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => [Node] -> m a
+  unmarshalNodes [x] = do
+    goto (nodeTSNode x)
+    to <$> gunmarshalNode
+  unmarshalNodes [] = fail "expected a node but didn't get one"
+  unmarshalNodes _ = fail "expected a node but got multiple"
 
 instance Unmarshal Text.Text where
-  unmarshalNode = do
-    node <- peekNode
-    case node of
-      Just node' -> do
-        bytestring <- ask
-        let start = fromIntegral (nodeStartByte node')
-            end = fromIntegral (nodeEndByte node')
-        pure (decodeUtf8 (slice start end bytestring))
-      _ -> fail "expected a node containing Text but didn't get one"
+  unmarshalNodes [node] = do
+    bytestring <- ask
+    let start = fromIntegral (nodeStartByte node)
+        end = fromIntegral (nodeEndByte node)
+    pure (decodeUtf8 (slice start end bytestring))
+  unmarshalNodes [] = fail "expected a node but didn't get one"
+  unmarshalNodes _ = fail "expected a node but got multiple"
 
 instance Unmarshal a => Unmarshal (Maybe a) where
-  unmarshalNode = Just <$> unmarshalNode
-  unmarshalEmpty = pure Nothing
+  unmarshalNodes [] = pure Nothing
+  unmarshalNodes listOfNodes = Just <$> unmarshalNodes listOfNodes
```

Comment:
This allows us to idiomatically handle a `Maybe` case, where we produce a `Nothing` if we encounter 0 elements in the list, and unmarshal a list of nodes  for one or more and return the unmarshaled nodes wrapped in a `Just`.